image-builder/image_build.py

```diff
@@ -0,0 +1,159 @@
+            {
+                "Authorization": f"Bearer {self.access_token}",
+                "Accept": "application/json",
+            }
+        )
+
+    def get_access_token(self) -> str:
+        """
+        curl -X POST -d grant_type=refresh_token \
+          -d client_id=rhsm-api -d refresh_token=$o \
+          https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token
+        :return: access_token
+        """
+        sso = requests.Session()
+        response = sso.post(
+            "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token",
+            data={
+                "grant_type": "refresh_token",
+                "client_id": "rhsm-api",
+                "refresh_token": offline_token,
+            },
+        )
+        output = response.json()
+        print("Output from obtaining the access token:")
+        pprint(output)
+        return output["access_token"]
+
+    def check_token(self):
+        response = requests.get(
+            "https://api.access.redhat.com/account/v1/user",
+            headers={"Authorization": f"Bearer {self.access_token}"},
+        )
+        json_output = response.json()
+        print(f"User info (Code: {response.status_code}): ")
+        pprint(json_output)
+
+    def touch_api(self):
+        """access readonly API endpoints - mostly to check that all is good"""
+        url = f"{image_builder_api_url}/openapi.json"
+        print(f"Image Builder API is available at {url}")
+        response = self.image_builder_session.get(url)
+        print(f"Response [{response.status_code}]")
+        pprint(response.json())
+
+        print("/ready version")
+        response = self.image_builder_session.get(f"{image_builder_api_url}/version")
+        print(f"Response [{response.status_code}]")
+        pprint(response.text)
+
+        print("/ready endpoint")
+        response = self.image_builder_session.get(f"{image_builder_api_url}/ready")
+        print(f"Response [{response.status_code}]")
+        pprint(response.text)
+
+    def create_image(self):
+        payload = {
+            "image_name": f"{os.getlogin()}-{time.time()}-90",
+            "distribution": "rhel-90",
+            "image_requests": [
+                {
+                    "architecture": "x86_64",
+                    "image_type": "aws",
+                    "upload_request": {
+                        "type": "aws",
+                        "options": {"share_with_accounts": ["727920394381"]},
+                    },
+                }
+            ],
+            "customizations": {
+                "packages": ["cockpit"],
+                "payload_repositories": [
+                    {
+                        "rhsm": False,
+                        "baseurl": (
+                            "https://download.copr.fedorainfracloud.org"
+                            "/results/@cockpit/cockpit-preview/centos-stream-9-x86_64/"
+                        ),
```

Comment:
Is it possible to be really explicit and use NVRs?

Reply:
I'll try to do that in the packages section to pick a specific build. It should work :)

Reply:
just confirmed this works

except that Copr only provides the latest build of package within a project

Reply:
> except that Copr only provides the latest build of package within a project

Not really, builds are removed no earlier than when they are 14 days old:
https://docs.pagure.org/copr.copr/user_documentation.html#how-long-do-you-keep-the-builds

DNF thought picks the latest-greatest NVR. If you care about particular PR-specific NVR, you
can install via fully specified NVR, like: "dnf install csdiff-2.5.0.20220615.104203.gde72e8f-1.fc36" (even thought the last version is csdiff-2.5.0.20220615.104203.gde72e8f-1.fc36.x86_64).

Reply:
ah, 2 weeks, so that's why the old cockpit build could not be found, thanks for correcting me!

Image Builder uses its own layer on top of dnf so it's possible they may need to explicitly account for this use case.